Fix site editor broken when fontWeight is not defined or is an integer in theme.json or theme styles

Imran92 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Fixes https://github.com/WordPress/gutenberg/issues/64952

First got report in https://github.com/Automattic/wp-calypso/issues/94033

## What?
<!-- In a few words, what is the PR actually doing? -->

In some themes (for example, Pendant), when users selected a style and then tried changing the font, the site editor crashed.

Upon digging, I found that it began after this https://github.com/WordPress/gutenberg/pull/61915. The PR added a very cool feature. Previously, we showed a fixed set of font weights in the editor, which was not always correct. This PR made it dynamic. But there were three small but tricky issues, all of them caused the site editor to stop working.

**1. trim() error**

If we look at [this line](https://github.com/WordPress/gutenberg/blob/f871c01ddcf27f20448da93737ef1aaf48c720d8/packages/block-editor/src/utils/get-font-styles-and-weights.js#L82), we'll notice that here ( `/\s/.test( face.fontWeight.trim()` ) the trim function will be called on the fontWeight variable even if it's undefined or integer, in which case it will throw an error.

Now, the question is, can fontWeight be undefined or integer? According to the `theme.json` [schema](https://raw.githubusercontent.com/WordPress/gutenberg/trunk/schemas/json/theme.json), we can see that it can be both. Because it's not required, and can be either a string or an integer (may change in later versions I think).

As real example, the Pendant theme doesn't have a `fontWeight` explicitly set [here](https://github.com/Automattic/themes/blob/dd40a6ef4033ad9f1649ea4bb7d2e6147c4346b9/pendant/styles/dark-navy.json#L64-L71).

<img width="602" alt="Screenshot 2024-08-31 at 3 08 33 AM" src="https://github.com/user-attachments/assets/8aaedaa9-7dd3-475f-b657-bc5955ab08df">

**2. Empty weight option**

In [this portion](https://github.com/WordPress/gutenberg/blob/f871c01ddcf27f20448da93737ef1aaf48c720d8/packages/block-editor/src/utils/get-font-styles-and-weights.js#L121-L131) a few lines below, we check if `fontWeight` value is present, and push it in a weight list if it does. But, we get the `fontWeight` value by calling the [formatFontWeight](https://github.com/WordPress/gutenberg/blob/5a87fb03b2f7cdcd56cee8d925349bf4c443a714/packages/block-editor/src/utils/format-font-weight.js#L12-L63) which returns an empty object even when `fontWeight` doesn't exist. So this check always passes. Resulting in listing an empty weight in the option.

**3. Infinite render loop and editor crash when weight is an integer**

This was the most tricky one as there was nothing helpful found in debug console, just broke the render depth after a while and crashed.

In this [line](https://github.com/WordPress/gutenberg/blob/f871c01ddcf27f20448da93737ef1aaf48c720d8/packages/block-editor/src/components/global-styles/typography-panel.js#L271-L274), you'll notice that we're trying to fetch the nearest weight value to set to the editor, by calling the `findNearestFontWeight` function. But if we look inside the function, we'll see that it returns an empty string if the `fontWeight` value of the type `string`. But we've already seen in the screenshot above that it's possible to have an integer as a fontWeight value. Pendant theme has a [few of those as well](https://github.com/Automattic/themes/blob/dd40a6ef4033ad9f1649ea4bb7d2e6147c4346b9/pendant/styles/dark-navy.json#L86-L116). So this function will not work if the fontWeight value is an integer. So it kept trying to get a value, till it caused it to crash.
https://github.com/WordPress/gutenberg/blob/5a87fb03b2f7cdcd56cee8d925349bf4c443a714/packages/block-editor/src/components/global-styles/typography-utils.js#L177-L179

Initially I thought about changing all the values as expected on the Pendant theme. But then decided it won't probably be the right thing to do because there can be other themes like this as well. Also, it follows the `theme.json` schema.

About the `'number' === typeof` approach that I took, instead of calling `toString()` on everything is because I noticed, even though it says fontWeight can only be an integer or a string, this [line](https://github.com/WordPress/gutenberg/blob/5a87fb03b2f7cdcd56cee8d925349bf4c443a714/packages/block-editor/src/utils/format-font-weight.js#L17-L19) indicates that fontWeight can sometimes be an object too (?). So I didn't want to risk breaking any backward compatibility.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

To fix the editor.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Having additional null checks and integer to string conversions.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
1. Any site
2. installed Pendant theme
3. Browse to Site Editor, then Styles. 
4. Select a style
5. Drilled down into Styles to Headings.
6. Select H1.
8. Select any other font
9. Editor shouldn't crash
10. Now try with different other themes as well, to make sure nothing broke.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

**Before**

https://github.com/user-attachments/assets/5a75bb70-71f3-4cdb-b9c5-ba6596579bdd

**After**

https://github.com/user-attachments/assets/074283f0-dcad-426e-a086-81146ea64bee

Imran92 committed:
Handle if nearest value input is not string for fontweight

Imran92 committed:
handle string and undefined fontweights when getting in site editor

ramonjd commented:
Thank your for this PR!

It's possible that this also fixes:

- https://github.com/WordPress/gutenberg/issues/44857

ramonjd commented on packages/block-editor/src/utils/get-font-styles-and-weights.js:
```diff
@@ -79,7 +79,10 @@ export function getFontStylesAndWeights( fontFamilyFaces ) {
-		if ( /\s/.test( face.fontWeight.trim() ) ) {
+		if (
+			'string' === typeof face.fontWeight &&
+			/\s/.test( face.fontWeight.trim() )
```
If we're guarding against extra whitespace for space-separated values, I'm wondering if we should trim all `face.fontWeight` string values, e.g., 

```js
		face.fontWeight =
			'string' === typeof face.fontWeight
				? face.fontWeight.trim()
				: face.fontWeight;
		if ( /\s/.test( face.fontWeight ) ) {
```

Imran92 replied:
Thanks so much for checking this and the suggestions @ramonjd ! I thought about writing tests for it right after adding the solution but then decided to put in the effort after we agreed on the solution's approach. As I'm very new to this area, there could be other ways to solve it which could be way better :p So wanted to make sure I'm doing it right.

`If we're guarding against extra whitespace for space-separated values, I'm wondering if we should trim all face.fontWeight`

This is a very nice idea. My only concern about this is that we'll probably be doing the regex test even on numeric values in the right next line.

Also, the regex test will return true if the value of fontweight variable is an object. It should probably never be an object, but this [line of here](https://github.com/WordPress/gutenberg/blob/5a87fb03b2f7cdcd56cee8d925349bf4c443a714/packages/block-editor/src/utils/format-font-weight.js#L17-L19) makes me wonder if it sometimes is an object for some reason.

ramonjd replied:
Ah yes, true. Thanks for the explanation 👍🏻 

> I thought about writing tests for it right after adding the solution but then decided to put in the effort after we agreed on the solution's approach.

Maybe tests will reveal some optimizations too. For now, it LGTM to me since it fixes a pretty major bug 🚀

ramonjd commented on packages/block-editor/src/utils/get-font-styles-and-weights.js:
```diff
@@ -105,7 +108,11 @@ export function getFontStylesAndWeights( fontFamilyFaces ) {
-		const fontWeight = formatFontWeight( face.fontWeight );
+		const fontWeight = formatFontWeight(
+			'number' === typeof face.fontWeight
+				? face.fontWeight.toString()
```
Could we coerce all values to string at the top?

I'm thinking, if we make sure all items in the `fontWeights` array are string here, at the source, the `toString()` coercions above in packages/block-editor/src/components/global-styles/typography-utils.js might not be required.

Imran92 replied:
This is a very good idea @ramonjd ! I actually tried to implement it. But later I stopped, because I was afraid of adding an regression because we're modifying the value fetched from the JSON. So for now, I'm just avoiding doing it in this PR. But WDYT about creating an issue where we properly test all implications of and modify the values at the top?

Thank you!

ramonjd replied:
Fair enough - probably a premature optimization. Good to have the bug fixed first 😄

ramonjd reviewed (commented):
Thank you for getting this started! I can replicate the error (`Maximum update depth exceeded`).

And this branch as it is fixes the problem for me.

I added some comments/questions.

I also think that it would be good to add test coverage too.

I did a bit of poking around in theme.json to see if we could coerce or warn about value types that don't conform to the schema definitions (related issue https://github.com/WordPress/gutenberg/issues/57641) but I think that's a separate issue. 

cc @mikachan / @creativecoder whose font library knowledge is deeper than mine

mikachan commented on packages/block-editor/src/utils/get-font-styles-and-weights.js:
```diff
@@ -118,7 +125,8 @@ export function getFontStylesAndWeights( fontFamilyFaces ) {
 		}
-		if ( fontWeight ) {
+
+		if ( Object.keys( fontWeight ).length ) {
```
I think this change is great. What do you think to also updating the `if ( fontStyle )` check, on line 119, so that these checks are consistent?

Imran92 replied:
Thanks @mikachan ! I've added the check here https://github.com/WordPress/gutenberg/pull/64953/commits/ca110e5d27710e7d850d4146e9f4133cc6d8ea9a

mikachan reviewed (commented):
Thanks for working on this, @Imran92! This fixes the issues you describe locally for me when testing with Pendant.

It would be great to update the tests for these functions before bringing this in, but otherwise, this is looking great.

Imran92 committed:
Add additional check for fontStyle keys

Imran92 committed:
Test getting styles and fonts for numeric and undefined fontweight

Imran92 committed:
Test numeric font style getting in typography utils

Imran92 committed:
Add test for nearest value when fontWeight undefined

Imran92 commented:
Thank you so much for checking it and your super helpful reviews @ramonjd and @mikachan! I've updated the PR as per the suggestions and have also added tests. Can you kindly take a look again?

mikachan reviewed (approved):
Thanks @Imran92!

The additional tests and changes look good to me; the tests are passing for me locally, I believe they cover the new changes, and the functionality is working well still in the Editor. 

I'm going to approve, but happy to wait for @ramonjd to take another look before we bring this in.

ramonjd reviewed (approved):
Thanks for fixing this, and for the tests.

ramonjd commented:
Also, thanks for the PR @Imran92 - I hope this is the first of many 🚀
